catalog/app/components/JsonEditor/Column.js

```diff
@@ -0,0 +1,161 @@
+import * as React from 'react'
+import isArray from 'lodash/isArray'
+import { useTable } from 'react-table'
+
+import * as M from '@material-ui/core'
+
+import AddArrayItem from './AddArrayItem'
+import AddRow from './AddRow'
+import Breadcrumbs from './Breadcrumbs'
+import Cell from './Cell'
+import Row from './Row'
+import { ColumnIds, EmptyValue } from './State'
+
+const useStyles = M.makeStyles((t) => ({
+  root: {
+    background: '#fff',
+    width: '100%',
+    flex: 'none',
+
+    '& + $root': {
+      marginLeft: -1,
+    },
+  },
+
+  tableContainer: {
+    padding: '1px 0', // NOTE: fit 2px border for input
+  },
+
+  selected: {
+    backgroundColor: t.palette.action.focus,
+  },
+}))
+
+function getColumntType({ parent, schema }) {
+  if (!parent) {
+    return schema.type
+  }
+
+  if (isArray(parent)) {
+    return 'array'
+  }
+
+  return typeof parent
+}
+
+export default function Table({
+  columnPath,
+  data,
+  onAddRow,
+  onCollapse,
+  onExpand,
+  onMenuAction,
+  updateMyData,
+}) {
+  const columns = React.useMemo(
+    () => [
+      {
+        accessor: ColumnIds.Key,
+      },
+      {
+        accessor: ColumnIds.Value,
+      },
+    ],
+    [],
+  )
+
+  const classes = useStyles()
+
+  const [hasNewRow, setHasNewRow] = React.useState(false)
+  const onChange = React.useCallback(
+    (...params) => {
+      setHasNewRow(false)
+      updateMyData(...params)
+    },
+    [updateMyData],
+  )
+
+  const tableInstance = useTable({
+    columns,
+    data: data.items,
+
+    defaultColumn: {
+      Cell,
+    },
+
+    updateMyData: onChange,
+  })
+  const { getTableProps, getTableBodyProps, rows, prepareRow } = tableInstance
+
+  const columnType = getColumntType(data)
+
+  const onAddRowInternal = React.useCallback(
+    (...params) => {
+      setHasNewRow(true)
+      onAddRow(...params)
+    },
+    [onAddRow],
+  )
+
+  return (
+    <div className={classes.root}>
+      {Boolean(columnPath.length) && (
```

Comment:
I prefer `Boolean(x)` instead of `!!x` for the reason it reads as more intentional. Comparing "I want not not some value" and "I want boolean representation of some value".

Reply:
it makes perfect sense, but i think `!!` is a widespread convention, almost an idiom, so i'm quite used to it